### PR TITLE
Router - calculating route/s which result multi-paths

### PR DESCRIPTION
**Pull request by saledjenic**
- `preferedChainIDs` removed from the input params for router v2
- typo in `disabledToChaindIDs` solved
- some improvements applied to the logic which generates candidates
- test for `resolveCandidates` added

Added tests:
- ETH transfer - No Specific FromChain - No Specific ToChain
- ETH transfer - No Specific FromChain - Specific Single ToChain
- ETH transfer - No Specific FromChain - Specific Multiple ToChain
- ETH transfer - Specific Single FromChain - No Specific ToChain
- ETH transfer - Specific Multiple FromChain - No Specific ToChain
- ETH transfer - Specific Single FromChain - Specific Single ToChain - Same Chains
- ETH transfer - Specific Single FromChain - Specific Single ToChain - Different Chains
- ETH transfer - Specific Multiple FromChain - Specific Multiple ToChain - Single Common Chain
- ETH transfer - Specific Multiple FromChain - Specific Multiple ToChain - Multiple Common Chains
- ETH transfer - Specific Multiple FromChain - Specific Multiple ToChain - No Common Chains
- ETH transfer - All FromChains Disabled - All ToChains Disabled
- ERC20 transfer - No Specific FromChain - No Specific ToChain
- ERC20 transfer - No Specific FromChain - Specific Single ToChain
- ERC20 transfer - No Specific FromChain - Specific Multiple ToChain
- ERC20 transfer - Specific Single FromChain - No Specific ToChain
- ERC20 transfer - Specific Multiple FromChain - No Specific ToChain
- ERC20 transfer - Specific Single FromChain - Specific Single ToChain - Same Chains
- ERC20 transfer - Specific Single FromChain - Specific Single ToChain - Different Chains
- ERC20 transfer - Specific Multiple FromChain - Specific Multiple ToChain - Single Common Chain
- ERC20 transfer - Specific Multiple FromChain - Specific Multiple ToChain - Multiple Common Chains
- ERC20 transfer - Specific Multiple FromChain - Specific Multiple ToChain - No Common Chains
- ERC20 transfer - All FromChains Disabled - All ToChains Disabled
- Bridge - No Specific FromChain - No Specific ToChain
- Bridge - No Specific FromChain - Specific Single ToChain
- Bridge - No Specific FromChain - Specific Multiple ToChain
- Bridge - Specific Single FromChain - No Specific ToChain
- Bridge - Specific Multiple FromChain - No Specific ToChain
- Bridge - Specific Single FromChain - Specific Single ToChain - Same Chains
- Bridge - Specific Single FromChain - Specific Single ToChain - Different Chains
- Bridge - Specific Multiple FromChain - Specific Multiple ToChain - Single Common Chain
- Bridge - Specific Multiple FromChain - Specific Multiple ToChain - Multiple Common Chains
- Bridge - Specific Multiple FromChain - Specific Multiple ToChain - No Common Chains
- Bridge - All FromChains Disabled - All ToChains Disabled
- ETH transfer - No Specific FromChain - No Specific ToChain - Single Chain LockedAmount
- ETH transfer - No Specific FromChain - No Specific ToChain - Multiple Chains LockedAmount
- ETH transfer - No Specific FromChain - No Specific ToChain - All Chains LockedAmount with insufficient amount
- ETH transfer - No Specific FromChain - No Specific ToChain - LockedAmount exceeds sending amount

### CODE REVIEW

File: services/wallet/router/pathprocessor/processor.go [hunk 1]
```diff
@@ -44,4 +44,9 @@ type ProcessorInputParams struct {
 	Username  string
 	PublicKey string
 	PackID    *big.Int
+
+	// for testing purposes
```
Comment (IvanBelyakoff) [inline services/wallet/router/pathprocessor/processor.go:59]: I think it is no good to place test fields in real types, especially when you further implement special logic for handling them in production code. We should separate those.
I believe we should replace it with mocking. For example:

- mock PathProcessor interface:
`mockgen -source=services/wallet/router/pathprocessor/processor.go -destination=services/wallet/router/pathprocessor/mock/processor.go -package=mock_pathprocessor`
- Instantiate mocked type in tests:
```
	mockCtrl := gomock.NewController(t)
	mockPathProcessor := mock_pathprocessor.NewMockPathProcessor(mockCtrl)
```
- Use expect calls:
`mockPathProcessor.EXPECT().EstimateGas(params[i]).Return(expectedGasFees[i], nil)`

File: services/wallet/router/pathprocessor/processor_bridge_celar.go [hunk 2]
```diff
@@ -229,6 +229,13 @@ func (c *CelerBridgeProcessor) PackTxInputData(params ProcessorInputParams) ([]b
 }
 
 func (s *CelerBridgeProcessor) EstimateGas(params ProcessorInputParams) (uint64, error) {
+	if params.TestsMode {
```
Comment (IvanBelyakoff) [inline services/wallet/router/pathprocessor/processor_bridge_celar.go:233]: As I mentioned above, we should not mix test-only code with production code

File: services/wallet/router/router_v2.go [hunk 3]
```diff
@@ -32,6 +32,33 @@ var (
 	}
 )
 
+const (
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go]: Better move it to tests as well

**Comment by saledjenic** [comment]
@IvanBelyakoff thanks for the early feedback, I agree, we have an issue separately for that https://github.com/status-im/status-desktop/issues/14849 but here I needed to test all that quickly to see if the parts I am checking are working well.

If I have time will do the tests properly within this PR, but at this point (since I am not too familiar with tests) this was faster for me and also not sure that I can do the tests I want without changing the code.

**Review by IvanBelyakoff — COMMENTED** [review]
Sale, I know it is a draft, but I am worried about some things, I left the comments inline.

File: services/wallet/router/router_v2.go [hunk 4]
```diff
@@ -377,15 +413,28 @@ func (r *Router) SuggestedRoutesV2(ctx context.Context, input *RouteInputParams)
...
+		networks  []*params.Network
+	)
+
+	networks, err = r.rpcClient.NetworkManager.Get(false)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		group = async.NewAtomicGroup(ctx)
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go:460]: I know it was like this before, but I believe candidates are independent of each other and if one candidate calculation fails, we should not stop calculation of the rest candidates, so I don't see a reason to use `AtomicGroup` here.
Just `async.Group` should better suite this case imo

Comment (saledjenic) [reply services/wallet/router/router_v2.go:460]: @IvanBelyakoff in my opinion `AtomicGroup` is a perfect candidate for this usage.

Have you checked in which cases we're returning an error?

Comment (IvanBelyakoff) [reply services/wallet/router/router_v2.go:460]: As far as I see, this line can return error:
`client, err := r.rpcClient.EthClient(network.ChainID)`
So each candidate is calculated for specific network. So if for some network candidate can't be found (due to some error) does it mean that other candidates should not be calculated and we should quit?

Comment (saledjenic) [reply services/wallet/router/router_v2.go:460]: Do we want to say this is the cheapest route even if we're not sure about it (cause we haven't checked all options)?

Also pretty sure if it fails for a single, will fail for all networks, cause the reason to have a failure is most likely "no internet" issue, also we're using the same provider for all chains.

Comment (IvanBelyakoff) [reply services/wallet/router/router_v2.go:460]: we can display more routes, but might be not the cheapest, or we dont display any routes at all or less routes. I prefer any routes rather than no routes. that lets user use those returned routes to make the transaction.

> Also pretty sure if it fails for a single, will fail for all networks, cause the reason to have a failure is most likely "no internet" issue, also we're using the same provider for all chains.

The reason for failure may be a coding bug anywhere - getting client from cache, getting network from DB, the passed chain should not be on the list. This is not a network-dependent issue. We don't do network requests here.

**Comment by Samyoul** [comment]
@saledjenic before I review would you like for me to wait until you've converted this PR from a draft to a full PR?

**Comment by saledjenic** [comment]
> @saledjenic before I review would you like for me to wait until you've converted this PR from a draft to a full PR?

Yes, that would be good, since the tests I added are more here to help me find the implementation issues. 
Will see at the end, once I finish all the improvements whether remove them and add them later via another issue or mock `PathProcessor`.

**Comment by saledjenic** [comment]
This PR is ready for review in terms of the correctness of the route/s calculation for bridge/transfer sending type which results in no, single, or multi-path for a single route.

Tests that you can see here helped me test and improve the router logic. 

I am not sure how I can achieve the same what I am testing with these tests if I would mock the `Router`?
If the tests are discutable here in the form they are, since they add some logic to the production code, I can remove them completely.

@alaibe @Samyoul @IvanBelyakoff

File: services/wallet/router/router_v2.go [hunk 5]
```diff
@@ -51,6 +50,22 @@ type RouteInputParams struct {
 	Username  string       `json:"username"`
 	PublicKey string       `json:"publicKey"`
 	PackID    *hexutil.Big `json:"packID"`
+
+	// Used for tests only
+	TestsMode  bool
+	TestParams *RouterTestParams
+}
+
+type RouterTestParams struct {
```
Comment (alaibe) [inline services/wallet/router/router_v2.go]: Is there any way we could have a test processor that we would inject, having an API like:

```
funct setProcessor() {
}
```
or
```
func setTestMode() {
}
```
, and instead of of doing
if testMode in code we would leverage the injected test element


Note: we could do same for fees and inject a test fees manager?

File: services/wallet/router/router_v2.go [hunk 6]
```diff
@@ -221,8 +236,13 @@ func findBestV2(routes [][]*PathV2, tokenPrice float64, nativeChainTokenPrice fl
...
 			tokenDenominator := big.NewFloat(math.Pow(10, float64(path.FromToken.Decimals)))
 
-			path.requiredTokenBalance = new(big.Int).Set(path.AmountIn.ToInt())
+			path.requiredTokenBalance = big.NewInt(0)
 			path.requiredNativeBalance = big.NewInt(0)
+			if path.FromToken.IsNative() {
+				path.requiredNativeBalance.Add(path.requiredNativeBalance, path.AmountIn.ToInt())
+			} else {
+				path.requiredTokenBalance.Add(path.requiredTokenBalance, path.AmountIn.ToInt())
+			}
```
Comment (alaibe) [inline services/wallet/router/router_v2.go:246]: why those changes?

Also I see this is duplicated , maybe we could do:

path.UpdateRequiredBalance()

File: services/wallet/router/router_v2.go [hunk 7]
```diff
@@ -294,98 +322,141 @@ func findBestV2(routes [][]*PathV2, tokenPrice float64, nativeChainTokenPrice fl
...
 		}
 
 		if input.AmountOut != nil && input.AmountOut.ToInt().Sign() < 0 {
-			return errors.New("amountOut must be positive")
+			return routererrors.ErrAmountOutMustBePositive
 		}
 	}
 
 	if input.FromLockedAmount != nil && len(input.FromLockedAmount) > 0 {
+		copyMap := func(src map[uint64]bool) map[uint64]bool {
```
Comment (alaibe) [inline services/wallet/router/router_v2.go]: maybe make it a real func rather than anonymous?

File: services/wallet/router/router_v2.go [hunk 8]
```diff
@@ -294,98 +322,141 @@ func findBestV2(routes [][]*PathV2, tokenPrice float64, nativeChainTokenPrice fl
...
+			totalLockedAmount = new(big.Int).Add(totalLockedAmount, amount.ToInt())
+
 			if amount == nil || amount.ToInt().Sign() < 0 {
-				return errors.New("locked amount must be positive")
+				return routererrors.ErrLockedAmountMustBePositive
 			}
 		}
+
+		if totalLockedAmount.Cmp(input.AmountIn.ToInt()) > 0 {
+			return routererrors.ErrLockedAmountExceedsTotalSendAmount
```
Comment (alaibe) [inline services/wallet/router/router_v2.go]: Can this be checked before? I mean in validation?

File: services/wallet/router/router_v2.go (hunk 5 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go]: Yes, this is something I was thinking of, but couldn't figure out what would be the best approach to achieve the same what I currently have for tests? I mean, I can mock all types that implement `PathProcessor` and that should work, but even then I need to have that `if testsMode` to properly set up processors, or I am not entirely sure how to do that, not sure if I can do that with making an interface for the router and then `mockgen`?

File: services/wallet/router/router_v2.go (hunk 6 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go:246]: > why those changes?

Because there is a bonder fee for the native token as well, not only for erc20 tokens.

> Also I see this is duplicated , maybe we could do...

It's not duplicated, it adds amount to different variables.

File: services/wallet/router/router_v2.go (hunk 7 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go]: For now it's used only here, we can move it out when there is a need to have/use it in another place. 
I was thinking of moving it to common.go, but it's not that general, cause it works with certain type only `map[uint64]bool`.

File: services/wallet/router/router_v2.go (hunk 8 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go]: It is already part of `validateInputData` function, which is the first one called.

File: services/wallet/router/router_v2.go [hunk 9]
```diff
@@ -294,98 +322,141 @@ func findBestV2(routes [][]*PathV2, tokenPrice float64, nativeChainTokenPrice fl
 func validateInputData(input *RouteInputParams) error {
```
Comment (IvanBelyakoff) [inline services/wallet/router/router_v2.go:323]: So if I pass this input:
```
&RouteInputParams {
  SendType: Transfer
}
```
This function will return nil which means it is valid. But it is not. So proceed with this input. But we should not trust client to send us proper data, so we should really validate it. Since data validation depends on specific pathprocessor and required fields are different for each path processor, we must:

- validate generic or common input data in this function and delegate checking specific required data to `processor.ValidatePathInputData()`
- test each processors's `ValidatePathInputData` with a unit test
- test common input data in a unit test for this function. Test that proper pathprocessor's ValidatePathInputData are called with EXPECT.

Option 2 would be to generate a huge amount of valid and invalid data. Lets say, if there are 20 fields in Input params, and each field can have 5 different values, than you have 5^20 combinations to check to really validate the input. Well, maybe you are not interested in all of them, but there are still a ton of them.

Otherwise, we just pretend to be testing, give it 2 good params, 2 invalid params and thinking we are covered.

So there should be some compromise, but I still think that compromise lies somewhere more in unit testing imo

**Review by alaibe — APPROVED** [review]

File: services/wallet/router/router_v2.go (hunk 9 above)
Comment (saledjenic) [reply services/wallet/router/router_v2.go:323]: @IvanBelyakoff tests for `validateInputData` will be covered in this PR which should be merged soon https://github.com/status-im/status-go/pull/5323 also if we see that anything is missed/overlooked there, we will add, but the purpose of it is to not let invalid input params to get to the resolveCandidates or any other function below it.

> Option 2 would be...

This sounds ok to me, we can implement that within this issue https://github.com/status-im/status-desktop/issues/14849 and remove all those `if params.TestsMode` checks.

Comment (IvanBelyakoff) [reply services/wallet/router/router_v2.go:323]: > This sounds ok to me

5^20 is like 95 quadrillion combinations. This does not look fine to me

Comment (IvanBelyakoff) [reply services/wallet/router/router_v2.go:323]: The PR you mentioned suffers the same issue that I described. My empty request will be considered valid. And it is not even tested there.

Comment (saledjenic) [reply services/wallet/router/router_v2.go:323]: > 5^20 is like 95 quadrillion combinations. This does not look fine to me

Well, if we take into consideration all possible tokens that exist now, it's way more... hahahahha

So, as said above `validateInputData` will have its own tests, so any data coming to functions below should be considered valid since those are not functions exposed via API.
What we really need to check is a combination of 3 different chains for `DisabledFromChainIDs` and `DisabledToChainIDs` for ETH and any erc20 token, let's say `USDC`, that's all, which is a waaaay less than those quadrillions you mentioned.

**Comment by saledjenic** [comment]
I added `// TODO: Remove two fields below once we implement a better solution for tests` to `RouteInputParams` type.

File: services/wallet/router/pathprocessor/processor.go [hunk 10]
```diff
@@ -44,4 +52,11 @@ type ProcessorInputParams struct {
...
+	TestsMode                 bool
+	TestEstimationMap         map[string]uint64   // [brifge-name, estimated-value]
+	TestBonderFeeMap          map[string]*big.Int // [token-symbol, bonder-fee]
+	TestApprovalGasEstimation uint64
+	TestApprovalL1Fee         uint64
 }
```
Comment (Samyoul) [inline services/wallet/router/pathprocessor/processor.go:65]: Does `TestsMode` refer to test network interactions or to go test configurations?

File: services/wallet/router/pathprocessor/constants.go [hunk 11]
```diff
@@ -13,10 +13,13 @@ var (
 
 const (
 	IncreaseEstimatedGasFactor = 1.1
+	SevenDaysInSeconds         = 604800
```
Comment (Samyoul) [inline services/wallet/router/pathprocessor/constants.go]: I know that this number is accurate, would you consider showing the calculation instead of the full number. It makes accidentally changing this const much more difficult to do.

I don't have a strong opinion on this though, so if you prefer the full number that is completely fine with me.

```suggestion
	SevenDaysInSeconds         = 60 * 60 * 24 * 7
```

File: services/wallet/router/pathprocessor/processor_bridge_celar.go [hunk 12]
```diff
@@ -66,6 +67,10 @@ func (s *CelerBridgeProcessor) Name() string {
 	return ProcessorBridgeCelerName
 }
 
+func (s *CelerBridgeProcessor) Clear() {
+	// no need for clearing prodTransferConfig and testTransferConfig cause they don't depend on input params
+}
+
```
Comment (Samyoul) [inline services/wallet/router/pathprocessor/processor_bridge_celar.go]: I notice that only two structs actually has logic in their `Clear()` interface method implementation. Is it possible that this function could be part of a separate interface?

Something like the below:

```go
// Define a separate interface for the Clear method
type Clearable interface {
    Clear()
}
```

Then in the logic you can call like below, meaning only the structs that need to call `Clear()` implement this function.

```go
    if clearable, ok := c.(Clearable); ok {
        clearable.Clear()
    }
```

File: services/wallet/router/pathprocessor/processor.go (hunk 10 above)
Comment (saledjenic) [reply services/wallet/router/pathprocessor/processor.go:65]: It's for the test configurations if that prop is set to true, it means we should use values set in the specific tests.

File: services/wallet/router/pathprocessor/processor_bridge_hop.go [hunk 13]
```diff
@@ -325,6 +354,25 @@ func (h *HopBridgeProcessor) BuildTransaction(sendArgs *MultipathProcessorTxArgs
...
+				BonderFee:               &bigint.BigInt{Int: val},
+				EstimatedRecieved:       &bigint.BigInt{Int: res},
+				Deadline:                time.Now().Add(SevenDaysInSeconds).Unix(),
+			}
+			h.bonderFee.Store(bonderKey, bonderFee)
+			return val, ZeroBigIntValue, nil
+		}
+		return nil, nil, routererrors.ErrNoBonderFeeFound
+	}
+
```
Comment (Samyoul) [inline services/wallet/router/pathprocessor/processor_bridge_hop.go]: This logic reads quite a lot like a mock interface implementation. If this is the case, I wonder if we could inject through composition test related logic in the go test files.

Or we could implement a function override only in the test context, something like below:

```go
// processor_bridge_hop_test.go
// +build test

package pathprocessor

import "fmt"

// Extend the main struct with test-only logic
type HopBridgeProcessorWithTest struct {
	HopBridgeProcessor
}

// Override the method with test-specific logic
func (h *HopBridgeProcessor) CalculateFees() {
	// The test logic
}
```

File: services/wallet/router/pathprocessor/processor_bridge_celar.go (hunk 12 above)
Comment (saledjenic) [reply services/wallet/router/pathprocessor/processor_bridge_celar.go]: Done, thanks!

File: services/wallet/router/router_v2.go (hunk 7 above)
Comment (Samyoul) [reply services/wallet/router/router_v2.go]: There is actually a function for this in `filter.go`

File: services/wallet/router/routererrors/errors.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+	if err == nil {
+		return nil
+	}
+	if errResp, ok := err.(*ErrorResponse); ok {
+		return errResp
+	}
+	return &ErrorResponse{
+		Code:    "0",
+		Details: err.Error(),
+	}
```
Comment (Samyoul) [inline services/wallet/router/routererrors/errors.go]: I think that it makes sense for this code to be in an independent package, that is not related to services or wallet.

File: services/wallet/router/routererrors/errors.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+var (
+	ErrUsernameAndPubKeyRequiredForENSRegister   = &ErrorResponse{Code: ErrorCode("R-001"), Details: "username and public key are required for ENSRegister"}
+	ErrOnlySTTSupportedForENSRegisterOnTestnet   = &ErrorResponse{Code: ErrorCode("R-002"), Details: "only STT is supported for ENSRegister on testnet"}
+	ErrOnlySTTSupportedForENSReleaseOnTestnet    = &ErrorResponse{Code: ErrorCode("R-003"), Details: "only STT is supported for ENSRelease on testnet"}
+	ErrUsernameRequiredForENSRelease             = &ErrorResponse{Code: ErrorCode("R-004"), Details: "username is required for ENSRelease"}
+	ErrUsernameAndPubKeyRequiredForENSSetPubKey  = &ErrorResponse{Code: ErrorCode("R-005"), Details: "username and public key are required for ENSSetPubKey"}
+	ErrPackIDRequiredForStickersBuy              = &ErrorResponse{Code: ErrorCode("R-006"), Details: "packID is required for StickersBuy"}
+	ErrToTokenIDRequiredForSwap                  = &ErrorResponse{Code: ErrorCode("R-007"), Details: "toTokenID is required for Swap"}
```
Comment (Samyoul) [inline services/wallet/router/routererrors/errors.go]: Very cool implementation 😉 , I wonder if these errors now could be within the `router` package.

**Review by Samyoul — APPROVED** [review]
Thank you for this PR and for looping me into these changes. I've left a few more comments, and thank you for addressing the comments that I have given earlier today.

**Comment by igor-sirotin** [comment]
Rebased to get fixes:
- https://github.com/status-im/status-go/pull/5397
- https://github.com/status-im/status-go/pull/5396

**Review by igor-sirotin — CHANGES_REQUESTED** [review]
Wallet is not my responsibility, and I don't want to confront of course. But I really think we should prevent things like `testMode` in our code. 

With such special flags you're basically not testing the real code. You're only testing the special test path. It's testing of tests, it doesn't make sense.

@saledjenic can you please clarify why do we need such things? Can we do it in any other way?

**Comment by saledjenic** [comment]
> Wallet is not my responsibility, and I don't want to confront of course. But I really think we should prevent things like `testMode` in our code.
> 
> @saledjenic can you please clarify why do we need such things? Can we do it in any other way?

As discussed above, I can add (and was adding) files that will inherit processor types, but in that case test coverage went lower.

**Comment by saledjenic** [comment]
Ad this moment I can either remove tests, if you don't like that approach (I don't like it either), but for testing what I'm testing I don't have other way now.

**Comment by igor-sirotin** [comment]
> ... but for testing what I'm testing I don't have other way now.

I might be wrong, but at a glance it seems that `ethClient` and `httpClient` are the only things that needs to be mocked for proper testing of the processors. 

- `ethClient` is already an interface and has a mock:
https://github.com/status-im/status-go/blob/bd816f1e290cb98ab7e0ce22725c78a801e25054/rpc/chain/mock/client/client.go#L353-L354

- `httpClient` can be mocked as well, but in a different way, e.g. we do it here:
https://github.com/status-im/status-go/blob/8c0e24dc260031a822604f7a2771f22daea79aa4/protocol/messenger_linkpreview_test.go#L364-L371

@saledjenic what do you think?

**Comment by saledjenic** [comment]
> @saledjenic what do you think?

Yes, @igor-sirotin that can work, but do you know how many different variant os `transactor.EXPECT().FunctionName(params).Return()`, where `FunctionName` should cover all functions used in all processors (we have 10 processors) in order to cover all input params that should be user as `params` of `FunctionName` and how many different responses I have to generate to have the Router logic covered?
It's measured in thousands.

**Comment by saledjenic** [comment]
Something like this could be a setup for that:
```
        transactor := mock_transactor.NewMockTransactorIface(mockCtrl)

	client, _ := rpc.NewClient(nil, 1, params.UpstreamRPCConfig{Enabled: false, URL: ""}, defaultNetworks, db)

	rpcClientEth := mock_client.NewMockClientInterface(mockCtrl)
	client.SetClient(mainnet.ChainID, rpcClientEth)
	rpcClientOpt := mock_client.NewMockClientInterface(mockCtrl)
	client.SetClient(optimism.ChainID, rpcClientOpt)
	rpcClientArb := mock_client.NewMockClientInterface(mockCtrl)
	client.SetClient(arbitrum.ChainID, rpcClientArb)

	router := NewRouter(client, transactor, nil, nil, nil, nil, nil, nil)
```

And then we're calling `routes, err := router.SuggestedRoutesV2(context.Background(), tt.input)`

And please have a look at how many different scenarios we currently have for input params of `SuggestedRoutesV2`, and please check the logic implemented in `SuggestedRoutesV2`, so if I need to mock all of them that's thousands of different cases.

**Comment by igor-sirotin** [comment]
> If the tests are discutable here in the form they are, since they add some logic to the production code, I can remove them completely.

> Ad this moment I can either remove tests, if you don't like that approach (I don't like it either)

Choosing between "no tests" and "tests with special flag", I'd go with the former.
We shouldn't introduce such code just to satisfy code climate. 

But I don't understand why we open a separate issue https://github.com/status-im/status-desktop/issues/14849 for writing tests. If everyone agree that `TestsMode` is a bad idea, and there's a plan to write proper tests, why not write them right here?

cc @saledjenic

**Comment by saledjenic** [comment]
> But I don't understand why we open a separate issue https://github.com/status-im/status-desktop/issues/14849 for writing tests. If everyone agree that TestsMode is a bad idea, and there's a plan to write proper tests, why not write them right here?

@igor-sirotin that issue is for adding more tests for other router-related things if we want to add them, but for the things that I'm testing if we want to test that we need thousands of mocked calls to cover all that.

I cannot add them now, here, cause I have more urgent things to do and I need changes done here for that. 

The only thing we can do is to implement an approach from [this comment](https://github.com/status-im/status-go/pull/5362#discussion_r1647319418), that will remove `if params.TestsMode {` from processors, but will bring 10 more files, and also won't remove it everywhere, cause it will remain in the approval required function and maybe some places more.

**Comment by saledjenic** [comment]
@igor-sirotin so you're more for no tests than this? But if so our logic can be broken later and it is not easy to catch it?

**Comment by igor-sirotin** [comment]
> I cannot add them now, here, cause I have more urgent things to do and I need changes done here for that.

Well I'm not the one to judge priorities.
 
> The only thing we can do is to implement an approach from [this comment](https://github.com/status-im/status-go/pull/5362#discussion_r1647319418), that will remove `if params.TestsMode {` from processors, but will bring 10 more files, and also won't remove it everywhere, cause it will remain in the approval required function and maybe some places more.

Yes, mocking processors sound good to me.

> so you're more for no tests than this? 

Yes.
I believe, when you don't have a good move, it doesn't mean you should do a bad one.

> But if so our logic can be broken later and it is not easy to catch it?

Yes, that's why we want proper tests.

-----

I guess we can ask other people to sort out the dispute 🙂 
cc @cammellos @alaibe @osmaczko @Samyoul

**Comment by saledjenic** [comment]
> Yes, mocking processors sound good to me.

It's not mocking (using gmock and have `mock_processor.NewMockPathProcessorInterface()` and so on), it's about inheriting the real processor and implementing necessary functions the way we need them for the test and then injecting those instances into the router as possible paths. So it doesn't mock it, but just avoids `if params.TestsMode {` in functions where we need that, but that condition remains in other places, so not that we will gain to much at the end.

**Comment by IvanBelyakoff** [comment]
> code

Sale, as for the approach of testing, I think it is different regarding mocking. We test each processor with UT and mocking ncessary things, for example to test Hop processor's `EstimateGas`, we do something like:

`mockEthClient.EXPECT().EstimateGas(<params_properly_passed>).Return(<proper_return_value_for_input_params>).`

So there are some 2-3 variations of happy and unhappy paths to consider HopProcessor more or less properly tested.
Then in router if we have a method that calls `HopeProcessor.EstimateGas`, we test it the same way:

`mockHopProcessor.EXPECT().EstimateGas(<params_properly_passed>).Return(<proper_return_value_for_input_params>)`

where we make sure the params passed to (for examplle) SuggestedRoutes match the params passed to `mockHopProcessor`, and the same for return values.
So it is important to keep in mind, that we are not testing that it should return some REAL values, we test the flow - proper calls are made and parameters are passed properly, return values properly returned.
So again, this is several different input data, but that covers the whole flow.
So with this strategy we cover the flow and a number of different input data needed to generate to test it is several multitude less than if test it like a blackbox, considering that the quality of testing stays the same.
That means of course:
- each pathprocessor must be tested separately. But that is what we want right?
- each huge method that does everything should be split and be highly cohesive. then it becomes easily testable. But that is how it should be, right?

If we introduce `testMode` flags now, nothing will stop me to use them as a pattern tomorrow for my PRs. But we don't wan this right?

**Comment by alaibe** [comment]
@saledjenic @igor-sirotin @IvanBelyakoff 
Perfect is the enemy of good

I believe this diff is better than current state, is it perfect, no. I myself is not a big fan of the way things are tested, but they are, and before they weren't. I consider this a step toward the right direction

There is also a few other improvements in code itself (error management / removing extra params...)

I'd like to emphasis that testing the router is a difficult job because of the potential 10s of call to infura/pricing API. Mocking wouldn't help here as you just end up testing the implementation.
If we are interested in isolating all the call and write unit test, at the end you don't test the router, you don't prevent bug and you increase coverage
If we are interested in finding bugs, the test input/output as in this PR is way to go.

What is our goal?

Ultimately to test this code we need to write more code like a fake processor / fake pricer and/or what Sale did which means we never gonna reach the threshold. At least in a iterative manner.

As said, this is a step toward the right direction, i don't see what's the problem with merging this PR.


Edited:
`If we introduce testMode flags now, nothing will stop me to use them as a pattern tomorrow for my PRs. But we don't wan this right?`

but nothing stop you to change the way it is tested if you have better strategy?

**Comment by saledjenic** [comment]
@IvanBelyakoff @igor-sirotin to me is not a problem at all to remove tests, but in my opinion, is better to have them, any tests, even inappropriately implemented, and improve them over time than not have them.

It's not a problem at all to add tests for testing each processor separately, but that doesn't test the router logic.

Could any of you create tests in a proper way, that are testing, the router implementation?

**Comment by IvanBelyakoff** [comment]
> @IvanBelyakoff @igor-sirotin to me is not a problem at all to remove tests, but in my opinion, is better to have them, any tests, even inappropriately implemented, and improve them over time than not have them.
> 
> It's not a problem at all to add tests for testing each processor separately, but that doesn't test the router logic.
> 
> Could any of you create tests in a proper way, that are testing, the router implementation?

Ok, I will prepare and example, give me some time please. Testing bridges btw can be left out of scope of the PR, no problems with that at all.

**Comment by saledjenic** [comment]
> Ok, I will prepare and example, give me some time please. Testing bridges btw can be left out of scope of the PR, no problems with that at all.

Looking forward to seeing it, thank you!

**Comment by IvanBelyakoff** [comment]
> > Ok, I will prepare and example, give me some time please. Testing bridges btw can be left out of scope of the PR, no problems with that at all.
> 
> Looking forward to seeing it, thank you!

Hey @saledjenic @alaibe @igor-sirotin I have made a PR  we agreed on based on this one. Please have a look.
This is my vision of it. Though within this timeframe it is quite hard to make it better, but I think it illustrates the alternative. Please read the description https://github.com/status-im/status-go/pull/5419

Let me know please if you find it useful.

**Comment by saledjenic** [comment]
@IvanBelyakoff I left a few comments there.

@igor-sirotin @IvanBelyakoff could we merge this one and you can keep working on the test improvements in via https://github.com/status-im/status-desktop/issues/14849 ?

**Comment by igor-sirotin** [comment]
@saledjenic sorry I'm overloaded today. Will reply in a few hours. Tomorrow morning worst case.

**Review by igor-sirotin — COMMENTED** [review]
@saledjenic @alaibe 

I feel like this discussion is stagnating. We've tried to give arguments and draft ideas, tried to guide the discussion into solution direction, but it doesn't seem to be happening. It's just getting both sides tired of the chat and not even closer to consensus.

I'm unlocking the PR, but I strongly disagree with `testsMode` approach.

@cammellos I'm also not blocking this PR from being force-merged without satisfying diff-coverage requirement.

**Comment by saledjenic** [comment]
@igor-sirotin I agree, let's merge it, since the changes are really important and urgent for us.
It won't remain on this, @IvanBelyakoff will most likely continue on what he started in https://github.com/status-im/status-go/pull/5419 and if that at the end tests the router logic the way we expect, we will get rid of `testsMode` and have proper tests soon.

Could you @cammellos merge it, please?

**Comment by saledjenic** [comment]
After the discussion we reached a consensus to merge this PR since:
- we have to reach the milestone goals and these changes are needed
- we have some tests for the router logic now, and in the past we didn't have any test

We all agreed on that having a check like `if testsMode` in the production code is not the best thing and that we should not propagate things like that further and that we need to replace it with proper tests sooner or later.

There is an issue for writing the router tests specifically logged here: https://github.com/status-im/status-desktop/issues/14849

And also there are some demo started by @IvanBelyakoff here https://github.com/status-im/status-go/pull/5419

As soon as we find a solution which will test the router logic as a black box without using `testsMode` we will replace the tests introduced in this PR.

Thank you all for participating in this and offline discussion.
@cammellos @alaibe @igor-sirotin @IvanBelyakoff
